README.md
# valuable
functional-style serialization tools


HISTORY.rst
Release history
---------------

development
+++++++++++

0.1.1 (2018-01-17)
++++++++++++++++++

- small fixes

0.1.0 (2018-01-16)
++++++++++++++++++

- initial release


tests/conftest.py
import typing as t
from datetime import datetime

import pytest

from valuable import load
from valuable.utils import identity


@pytest.fixture
def registry():
    return load.PrimitiveRegistry({
        int:        int,
        float:      float,
        bool:       {'true': True, 'false': False}.__getitem__,
        str:        str,
        type(None): identity,
        object:     identity,
        datetime:   load.parse_iso8601,
    }) | load.GenericRegistry({
        t.List:  load.list_loader,
        t.Set:   load.set_loader,
    }) | load.get_optional_loader


tests/test_xml.py
import xml.etree.ElementTree as ET

import pytest

from valuable import xml


XMLDATA = ET.fromstring('''
<MyRoot foo="3">
    <MyParent>
        <Child1 bla="qux">foo</Child1>
        <Child1> bar  </Child1>
    </MyParent>
</MyRoot>
''')


class TestElemGetter:

    def test_simple(self):
        getter = xml.elemgetter('MyParent/Child1')
        elem = getter(XMLDATA)
        assert elem.text == 'foo'
        assert elem.attrib == {'bla': 'qux'}

    def test_not_found(self):
        getter = xml.elemgetter('MyParent/BlaBla')
        with pytest.raises(LookupError, match='MyParent/BlaBla'):
            getter(XMLDATA)


class TestElemsGetter:

    def test_simple(self):
        getter = xml.elemsgetter('MyParent/Child1')
        elems = getter(XMLDATA)
        assert len(elems) == 2
        assert [e.text for e in elems] == ['foo', ' bar  ']

    def test_not_found(self):
        getter = xml.elemsgetter('MyParent/Child2')
        assert getter(XMLDATA) == []


class TestTextGetter:

    def test_simple(self):
        getter = xml.textgetter('MyParent/Child1')
        assert getter(XMLDATA) == 'foo'

    def test_not_found(self):
        getter = xml.textgetter('MyParent/Child2')
        with pytest.raises(LookupError, match='MyParent/Child2'):
            getter(XMLDATA)

    def test_default(self):
        getter = xml.textgetter('MyParent/Child2', default=None)
        assert getter(XMLDATA) is None

        getter = xml.textgetter('MyParent/Child1', default=None)
        assert getter(XMLDATA) == 'foo'

    def test_strip(self):
        getter = xml.textgetter('MyParent/Child1[2]', strip=True)
        assert getter(XMLDATA) == 'bar'

    def test_strip_and_default(self):
        getter = xml.textgetter('MyParent/Child2', default=None, strip=True)
        assert getter(XMLDATA) is None


class TestAttribGetter:

    def test_simple(self):
        getter = xml.attribgetter('MyParent/Child1', 'bla')
        assert getter(XMLDATA) == 'qux'

    def test_path_not_found(self):
        getter = xml.attribgetter('MyParent/Child2', 'blabla')
        with pytest.raises(LookupError, match='MyParent/Child2'):
            getter(XMLDATA)

    def test_root_path(self):
        getter = xml.attribgetter('.', 'foo')
        assert getter(XMLDATA) == '3'

    def test_attrib_not_found(self):
        getter = xml.attribgetter('MyParent/Child1', 'foo')
        with pytest.raises(LookupError, match='foo'):
            getter(XMLDATA)

    def test_default(self):
        getter = xml.attribgetter('MyParent/Child1', 'foo', default=None)
        assert getter(XMLDATA) is None

        getter2 = xml.attribgetter('MyParent/Child1', 'bla', default=None)
        assert getter2(XMLDATA) == 'qux'


class TestTextsGetter:

    def test_simple(self):
        getter = xml.textsgetter('MyParent/Child1')
        assert getter(XMLDATA) == ['foo', ' bar  ']

    def test_path_not_found(self):
        getter = xml.textsgetter('MyParent/Child2')
        assert getter(XMLDATA) == []

    def test_strip(self):
        getter = xml.textsgetter('MyParent/Child1', strip=True)
        assert getter(XMLDATA) == ['foo', 'bar']


tests/test_load.py
import typing as t
from datetime import datetime, timedelta, timezone
from operator import itemgetter

import pytest

from valuable import load
from valuable.utils import compose


def dataclass_or_skip(*args, **kwargs):
    """create a dataclass or skip the test"""
    dataclasses = pytest.importorskip('dataclasses')
    return dataclasses.make_dataclass(*args, **kwargs)


def test_list_loader():
    assert load.list_loader((int, ), [4, '3', '-1']) == [4, 3, -1]


def test_set_loader():
    assert load.set_loader((int, ), [4, '3', '-1']) == {4, 3, -1}


def test_lookup_default():
    getter = load.lookup_defaults(itemgetter('foo'), 'bla')
    assert getter({}) == 'bla'
    assert getter({'foo': 4}) == 4


class TestPrimitiveRegistry:

    def test_found(self):
        registry = load.PrimitiveRegistry({int: round})
        loader = registry(int)
        assert loader(3.4) == 3

    def test_not_found(self):
        registry = load.PrimitiveRegistry({int: round})
        with pytest.raises(load.UnsupportedType):
            registry(str)


class TestGenericRegistry:

    def test_ok(self):

        Tag = t.NamedTuple('Tag', [('name', str)])

        registry = load.GenericRegistry({
            t.List: load.list_loader,
        }) | load.PrimitiveRegistry({
            Tag: compose(Tag, '<{}>'.format),
        })

        # simple case
        loader = registry(t.List[Tag])
        assert loader(['hello', 5, 'there']) == [
            Tag('<hello>'), Tag('<5>'), Tag('<there>')
        ]

        # recursive case
        loader = registry(t.List[t.List[Tag]])
        assert loader([
            ['hello', 9, 'there'],
            [],
            ['another', 'list']
        ]) == [
            [Tag('<hello>'), Tag('<9>'), Tag('<there>')],
            [],
            [Tag('<another>'), Tag('<list>')]
        ]

    def test_unsupported_type(self):

        Tag = t.NamedTuple('Tag', [('name', str)])

        registry = load.GenericRegistry({
            t.List: load.list_loader,
        }) | load.PrimitiveRegistry({
            Tag: compose(Tag, '<{}>'.format),
        })
        with pytest.raises(load.UnsupportedType):
            registry(t.Set[Tag])

        with pytest.raises(load.UnsupportedType):
            registry(t.List[str])

        with pytest.raises(load.UnsupportedType):
            registry(object)


class TestGetOptionalLoader:

    def test_ok(self):

        Tag = t.NamedTuple('Tag', [('name', str)])

        registry = load.PrimitiveRegistry({
            Tag: compose(Tag, '<{}>'.format)
        }) | load.get_optional_loader

        loader = registry(t.Optional[Tag])
        assert loader(None) is None
        assert loader(5) == Tag('<5>')


class TestParseIso8601:

    def test_with_timezone(self):
        parsed = load.parse_iso8601('2012-02-27T13:08:00+0100')
        assert parsed == datetime(
            2012, 2, 27, 13, 8,
            tzinfo=timezone(timedelta(hours=1)))

    def test_no_timezone(self):
        parsed = load.parse_iso8601('2014-06-10T17:25:29Z')
        assert parsed == datetime(2014, 6, 10, 17, 25, 29)


def test_dataclass_registry(registry):

    Post = dataclass_or_skip('Post', [
        ('title', str),
        ('posted_at', datetime),
        ('author_id', int)
    ])

    data = {
        'Title':     'hello',
        'date':      '2017-10-18T14:13:05Z',
        'author_id': 12,
    }

    registry |= load.DataclassRegistry({
        Post: {
            'title':     itemgetter('Title'),
            'posted_at': itemgetter('date'),
            'author_id': itemgetter('author_id'),
        }
    })

    loader = registry(Post)

    assert loader(data) == Post(
        'hello',
        datetime(2017, 10, 18, 14, 13, 5),
        author_id=12)

    Foo = dataclass_or_skip('Foo', [])

    with pytest.raises(load.UnsupportedType):
        registry(Foo)


def test_create_dataclass_loader(registry):
    User = dataclass_or_skip('User', [
        ('id', int),
        ('name', str),
        ('hobbies', t.Optional[t.List[str]]),
        ('nicknames', t.Set[str]),
        ('height', float),
        ('likes_pizza', t.Optional[bool], True),
    ])

    dloader = load.create_dataclass_loader(User, registry, {
        'id':          itemgetter('id'),
        'name':        itemgetter('username'),
        'likes_pizza': itemgetter('pizza'),
        'height':      itemgetter('height'),
        'nicknames':   itemgetter('nicknames'),
        'hobbies':     itemgetter('hobbies'),
    })
    assert dloader({
        'id': 98,
        'username': 'wilma',
        'hobbies': ['tennis', 'diving'],
        'pizza': 'true',
        'height': 1.64,
        'nicknames': []
    }) == User(98, 'wilma', hobbies=['tennis', 'diving'],
               likes_pizza=True, height=1.64, nicknames=set())


class TestAutoDataclassRegistry:

    def test_ok(self, registry):

        Post = dataclass_or_skip('Post', [
            ('title', str),
            ('posted_at', datetime),
        ])

        registry |= load.AutoDataclassRegistry()
        loader = registry(Post)

        data = {
            'title':     'hello',
            'posted_at': '2017-10-18T14:13:05Z'
        }
        assert loader(data) == Post(
            'hello',
            datetime(2017, 10, 18, 14, 13, 5))

    def test_not_supported(self, registry):

        class MyClass():
            pass

        registry |= load.AutoDataclassRegistry()

        with pytest.raises(load.UnsupportedType):
            registry(MyClass)


def test_simple_registry():

    User = dataclass_or_skip('User', [
        ('name', str),
        ('id', int),
        ('nickname', t.Optional[str]),
    ])

    Post = dataclass_or_skip('Post', [
        ('title', str),
        ('user', User),
        ('comments', t.List[str])
    ])

    loader = load.simple_registry(Post)

    loaded = loader({
        'title': 'hello',
        'comments': ['first!', 'another comment', 5],
        'user': {
            'name': 'bob',
            'nickname': 'bobby',
            'id': '543',
            'extra data': '...',
        }
    })
    assert loaded == Post(
        title='hello',
        comments=['first!', 'another comment', '5'],
        user=User(
            name='bob',
            id=543,
            nickname='bobby'
        )
    )


valuable/utils.py
"""Miscellaneous tools, boilerplate, and shortcuts"""
import typing as t
from collections import Mapping

T = t.TypeVar('T')


class NO_DEFAULT:
    """sentinel for no default"""


def identity(obj):
    """identity function, returns input unmodified"""
    return obj


class compose:
    """compose a function from a chain of functions

    Parameters
    ----------
    *funcs
        callables to compose

    Note
    ----
    * if given no functions, acts as :func:`identity`
    """
    def __init__(self, *funcs: t.Callable):
        self.funcs = funcs

    def __hash__(self):
        return hash(self.funcs)

    def __eq__(self, other):
        if isinstance(other, compose):
            return self.funcs == other.funcs
        return NotImplemented

    def __ne__(self, other):
        if isinstance(other, compose):
            return self.funcs != other.funcs
        return NotImplemented

    def __call__(self, *args, **kwargs):
        if not self.funcs:
            return identity(*args, **kwargs)
        *tail, head = self.funcs
        value = head(*args, **kwargs)
        for func in reversed(tail):
            value = func(value)
        return value


class _EmptyMapping(Mapping):
    """an empty mapping to use as a default value"""
    def __iter__(self):
        yield from ()

    def __getitem__(self, key):
        raise KeyError(key)

    def __len__(self):
        return 0

    def __repr__(self):
        return '{}'


EMPTY_MAPPING = _EmptyMapping()


valuable/load.py
"""Tools for deserialization

Todo
----
* resolve forward declarations in dataclass signatures
* auto namedtuple registry
"""
import abc
import typing as t
from datetime import datetime
from functools import partial
from itertools import starmap
from operator import attrgetter, itemgetter

from .utils import EMPTY_MAPPING, compose, identity

__all__ = ['Registry', 'Loader', 'CombinableRegistry', 'MultiRegistry',
           'PrimitiveRegistry', 'GenericRegistry', 'AutoDataclassRegistry',
           'create_dataclass_loader', 'list_loader', 'set_loader',
           'UnsupportedType', 'DataclassRegistry', 'lookup_defaults',
           'get_optional_loader', 'simple_registry', 'parse_iso8601']

T = t.TypeVar('T')
_NoneType = type(None)


class lookup_defaults:
    """wrap a lookup function with a default if lookup fails"""
    def __init__(self, lookup: t.Callable[[t.Any], T], default: T):
        self.lookup, self.default = lookup, default

    def __call__(self, obj):
        try:
            return self.lookup(obj)
        except LookupError:
            return self.default


def parse_iso8601(dtstring: str) -> datetime:
    """naive parser for ISO8061 datetime strings,

    Parameters
    ----------
    dtstring
        the datetime as string in one of two formats:

        * ``2017-11-20T07:16:29+0000``
        * ``2017-11-20T07:16:29Z``

    """
    return datetime.strptime(
        dtstring,
        '%Y-%m-%dT%H:%M:%SZ' if len(dtstring) == 20 else '%Y-%m-%dT%H:%M:%S%z')


class Loader(t.Generic[T]):
    """Interface for loaders (deserializers).
    Any callable with the signature ``Any -> T`` implements ``Loader[T]``
    """
    def __call__(self, value: t.Any) -> T:
        """loads an object of a particular type from any data"""
        raise NotImplementedError()


class Registry(abc.ABC):
    """Interface for a loader registry.
    Any callable with signature ``Type[T] -> Loader[T]`` implements it.
    """

    @abc.abstractmethod
    def __call__(self, cls: t.Type[T]) -> Loader[T]:
        raise NotImplementedError()


class CombinableRegistry(Registry):
    """base class for registries which may be combined

    any callable implemeting the signature
    ``(t.Type[T], Registry) -> Loader[T]``
    is combinable

    also provides ``__or__`` as a mixin method
    """

    @abc.abstractmethod
    def __call__(self, cls: t.Type[T], main: Registry) -> Loader[T]:
        raise NotImplementedError()

    def __or__(self, other: 'CombinableRegistry') -> 'MultiRegistry':
        return MultiRegistry(self, other)


class MultiRegistry(CombinableRegistry):
    def __init__(self, *children: CombinableRegistry):
        self.children = children

    def __call__(self, cls, main=None):
        exc = None
        for registry in self.children:
            exc = None
            try:
                return registry(cls, main=main or self)
            except UnsupportedType as excep:
                exc = excep
        raise exc

    def __or__(self, other: CombinableRegistry) -> 'MultiRegistry':
        return MultiRegistry(*self.children, other)


class PrimitiveRegistry(CombinableRegistry):
    """a registry of primitive (i.e. non-nested) loaders"""
    def __init__(self, registry=EMPTY_MAPPING):
        self.registry = registry

    def __call__(self, cls, main=None):
        try:
            return self.registry[cls]
        except KeyError:
            raise UnsupportedType(cls)


class GenericRegistry(CombinableRegistry):
    """registry for generic types. for example :class:`~typing.List`.

    These types must have ``__origin__`` and ``__args__`` attributes
    """
    def __init__(self, registry=EMPTY_MAPPING):
        self.registry = registry

    def __call__(self, cls: t.Type[T], main: Registry=None) -> T:
        if not hasattr(cls, '__origin__'):
            raise UnsupportedType(cls)
        try:
            return partial(self.registry[cls.__origin__],
                           list(map(main or self, cls.__args__)))
        except KeyError:
            raise UnsupportedType(cls)


class UnsupportedType(LookupError):
    """indicates the registry does not have a loader for the given type"""


def create_dataclass_loader(cls, registry, field_getters):
    """create a loader for a dataclass type"""
    fields = cls.__dataclass_fields__
    item_loaders = map(registry, map(attrgetter('type'), fields.values()))
    getters = map(field_getters.__getitem__, fields)
    loaders = list(starmap(compose, zip(item_loaders, getters)))

    def dloader(obj):
        return cls(*(g(obj) for g in loaders))

    return dloader


def list_loader(subloaders, value):
    """loader for the List generic"""
    loader, = subloaders
    return list(map(loader, value))


def set_loader(subloaders, value):
    """loader for the Set generic"""
    loader, = subloaders
    return set(map(loader, value))


if hasattr(t.Union, '__origin__'):
    def get_optional_loader(cls: t.Type[T], main: Registry) -> Loader[T]:
        """a combinable registry for optional types"""
        if _is_optional_type(cls):
            return partial(_optional_loader, main(cls.__args__[0]))
        else:
            raise UnsupportedType(cls)
else:  # pragma: no cover
    def get_optional_loader(cls: t.Type[T], main: Registry) -> Loader[T]:
        """a combinable registry for optional types"""
        if _is_optional_type(cls):
            return partial(_optional_loader, main(cls.__union_params__[0]))
        else:
            raise UnsupportedType(cls)


def _optional_loader(subloader, value):
    return value if value is None else subloader(value)


class AutoDataclassRegistry(CombinableRegistry):
    """registry which creates dataclass loaders on-the-fly"""

    def __call__(self, cls, main):
        if hasattr(cls, '__dataclass_fields__'):
            return create_dataclass_loader(cls, main, {
                f: itemgetter(f) for f in cls.__dataclass_fields__
            })
        else:
            raise UnsupportedType(cls)


class DataclassRegistry(CombinableRegistry):
    """registry for dataclasses"""
    def __init__(self, confs):
        self.confs = confs

    def __call__(self, cls, main):
        try:
            sourcemap = self.confs[cls]
        except KeyError:
            raise UnsupportedType(cls)
        return create_dataclass_loader(cls, main, sourcemap)


simple_registry = PrimitiveRegistry({
    int:        int,
    float:      float,
    str:        str,
    bool:       bool,
    type(None): identity,
    datetime:   parse_iso8601,
    object:     identity,
}) | GenericRegistry({
    t.List:   list_loader,
    t.Set:    set_loader,
}) | get_optional_loader | AutoDataclassRegistry()


if hasattr(t.Union, '__origin__'):
    def _is_optional_type(cls):
        """determine whether a class is an Optional[...] type"""
        try:
            return (cls.__origin__ is t.Union
                    and len(cls.__args__) == 2
                    and cls.__args__[1] is _NoneType)
        except AttributeError:
            return False
else:  # pragma: no cover
    def _is_optional_type(cls):
        """determine whether a class is an Optional[...] type"""
        try:
            return (len(cls.__union_params__) == 2
                    and cls.__union_params__[1] is _NoneType)
        except AttributeError:
            return False
